go/mysql/conn.go

```diff
@@ -898,11 +898,8 @@ func (c *Conn) handleNextCommand(handler Handler) error {
 		if stmtID != uint32(0) {
 			defer func() {
 				prepare := c.PrepareData[stmtID]
-				if prepare.BindVars != nil {
-					for k := range prepare.BindVars {
-						prepare.BindVars[k] = nil
-					}
-				}
```

Comment:
Though I agree that this is a cleaner and safer way of resetting PrepareData, I'm concerned that the fix might mask a deeper problem. Is it possible that there is a mismatch between ParamsCount and BindVars? Maybe we are being told to expect a certain number of bind variables (ParamsCount) but receiving a fewer number.

Reply:
ok, so should we mask and report errors or something? or would rather have things immediately fail?

Reply:
@deepthi That's a good point. Maybe the problem is that we are reusing `stmtID` values for different statements?

@PrismaPhonic Were you able to reproduce the problem with the reduced test case?

Reply:
Also, although this is cleaner and safer than setting everything to nil, there is a performance cost to that. It was probably written this way to avoid allocating a new map on every query. This "fix" will put more load on the GC, which is something we shouldn't do in the hot path unless there's a very good reason.

Reply:
@deepthi will as far as I know we set the number of parameters when we check the `:v` prefix and send it back to the client. Therefore, if there was a mismatch I think the client might catch that.

@enisoc I dont know of a situation that will change the query and number of parameters for a given s`stmtID`.

Reply:
OK this finally makes sense. The clue about `vtg1` in https://github.com/vitessio/vitess/pull/5441#issuecomment-555260146 was the key to finding the root cause.

The `vtg1` entry in the map is getting added by the normalizer after we pass the map to `VTGate.Execute()`. The way to reproduce this is to prepare a query that also has some raw values, such as the `'blah'` in `select foo from bar where name in (?,'blah')`. The normalizer will replace the `'blah'` with `:vtg1` and add `vtg1` to the bindvars map with value `'blah'`.

After executing, we set all bindvar map values to `nil` without deleting them, which would have been fine if the normalizer had not mutated our map in-place to add `vtg1`. Instead, the next time we call `VTGate.Execute()`, we pass a bindvars map with `vtg1` already present (and set to `nil`). The normalizer sees this and thinks you have a pre-existing bindvar called `vtg1` so when it needs to generate a bindvar name for `'blah'`, it increments to `vtg2` to avoid a collision. Now `vtg1` remains set in the map with value `nil`.

The fix for that is what this PR is already doing: make a new map every time. The only suggestion I have then is to add a comment explaining why this is necessary:

```go
// Allocate a new bindvar map every time since VTGate.Execute() mutates it.
```

Reply:
Added the comment with https://github.com/vitessio/vitess/pull/5441/commits/4411cda3fe0afa3332b1810c930ef3f52a71ab97.